IoT-Projects/ESP8266/NeoPixel Smart Planter/UBIDOTS_ESP8266_RGB_SMART_PLANTER.ino

```diff
@@ -0,0 +1,317 @@
+
+Made by: Maria Hernández - IoT Developer Advocate @ Ubidots
+Revision: José García - Development & Support Manager @ Ubidots
+
+/****************************************
+ * Include Libraries
+ ****************************************/
+#include <Adafruit_NeoPixel.h>
+#include <stdio.h>
+#include <map>
+#include "DHT.h"
+#include "UbidotsESPMQTT.h"
+
+/****************************************
+ * Define Pins
+ ****************************************/
+#define LIGHTPIN D1     // Digital pin for Led Lamp.
+#define DHTPIN D5       // Digital pin for DHT sensor.
+#define NEOPIXELSPIN D6 // Digital pin for NeoPixel Ring.
+#define MOISTUREPIN A0  // Analog pin for Moisture Sensor.
+
+/****************************************
+ * Define Constants
+ ****************************************/
+#define TOKEN "BBFF-xxxxxxxxxx" // Assign your Ubidots TOKEN.
+#define WIFINAME "xxxxxxxxxx"   // Assign your SSID.
+#define WIFIPASS "xxxxxxxxxx"   // Assign your WiFi Password.
+#define DEVICE "planter"        // Ubidots Device Label.
+#define VAR_PUB_1 "temperature" // Ubidots Variables' label for publishing data.
+#define VAR_PUB_2 "humidity"
+#define VAR_PUB_3 "soil-moisture"
+#define VAR_PUB_4 "heat-index"
+#define VAR_SUB_1 "light-1" // Ubidots Variables' label for subscribing to data; \
+                            // These variables have to be created at Ubidots.
+#define VAR_SUB_2 "light-2"
+#define NUMPIXELS 12 // 12 bit NeoPixel Ring
+// Uncomment whatever type you're using
+#define DHTTYPE DHT11 // DHT 11
+//#define DHTTYPE DHT22   // DHT 22  (AM2302), AM2321
+//#define DHTTYPE DHT21   // DHT 21 (AM2301)
+
+typedef enum
+{
+  red,
+  green,
+  blue,
+  yellow,
+  white,
+  black
+} NeoPixelColor;
+
+//   R,   G,   B
+uint8_t myColors[][6] = {{250, 0, 0},                             // Red.
+                         {0, 255, 0},                             // Green.
+                         {0, 0, 255},                             // Blue.
+                         {255, 255, 0},                           // Yellow.
+                         {255, 255, 255},                         // White.
+                         {0, 0, 0}};                              // Black.
+const uint8_t numberOfVariables = 2;                              // Number of variables for subscription.
+char *variableLabels[numberOfVariables] = {VAR_SUB_1, VAR_SUB_2}; // Variables' label for subscription.
+float value;                                                      // Store incoming value.
+int lastValue;                                                    // Store incoming value.
+bool bottomLight;                                                 // flag to control conditions for the bottom light.
+unsigned long initTime;                                           // Store the init time.
+const long SECONDS_TO_RECONNECT = 180000;                         // Period to reconnect MQTT connection.
+
+// Comparison functor to map functions.
+struct cmp_str
```

Comment:
where do you use this structure? from what I see you compare two char arrays, but I do not get where and why. From my review, the code should work without this variable

Reply:
The code doesn't work without this struct. As you can see below, this structure is used on the map declaration: 

typedef std::map<const char *, FunctionType, cmp_str> mapTopicSubscription;

This struct has to be defined anytime you want a map key to be a pointer. You must give a comparison functor to the map otherwise it's comparing the pointer, not the null-terminated string it points to.